src/Workspaces/Core/Portable/Execution/SerializerService_Reference.cs

```diff
@@ -58,6 +58,10 @@ public static Checksum CreateChecksum(AnalyzerReference reference, CancellationT
                     default:
                         throw ExceptionUtilities.UnexpectedValue(reference);
                 }
+
+                // We aren't going to deserialize the data from this stream, so no need to keep referenced objects alive
+                // longer than the serialization code.
+                _ = writer.TakeKeepAliveObjects();
```

Comment:
confused. then why pass `canKeepObjectsAlive: true`?

Reply:
(applies to all cases).

Reply:
This is preservation of existing behavior. It's also probably not what we want. Currently the checksum is being calculated against the _location of reference metadata in temporary storage_ as opposed to being calculated against the _content_.

Reply:
> confused. then why pass canKeepObjectsAlive: true

As a direct answer to this: This code can ensure that keep alive objects are kept alive until such time as the last ObjectReader operating on the serialized data has completed its serialization process. For checksum calculation, there are zero ObjectReader instances operating on the serialized data, so this condition is maintained regardless of how/when the keep alive objects are kept alive.

Reply:
➡️ Now documented on new property `Checksum.ChecksumKeepAliveCallback`